pkg/ovirt/ovclient/ovclient.go

```diff
@@ -0,0 +1,174 @@
+	// validate the secret that has the oVirt REST creds
+	creds, err := getCredentials(cli)
+	if err != nil {
+		return nil, err
+	}
+
+	creds.CA = ca
+
+	// Get an oVirt client
+	ovcli, err := ensureOvClient(creds, apiServerURL)
+	if err != nil {
+		return nil, err
+	}
+
+	return ovcli, nil
+}
+
+// ensureOvClient ensures that we have an access token that can be used with the oVirt REST API.
+func ensureOvClient(creds *Credentials, apiServerURL string) (*Client, error) {
+	// Create a new client and validate that it works
+	ovcli := &Client{
+		ApiServerURL: apiServerURL,
+		Credentials:  creds,
+	}
+
+	if err := ovcli.ensureAccessToken(); err != nil {
+		return nil, err
+	}
+
+	// Validate the token by geting system information
+	const path = "/api"
+
+	// call the server to get the datacenters just to verify the REST API works.
+	body, err := ovcli.REST.Get(ovcli.AccessToken, path)
+	if err != nil {
+		err = fmt.Errorf("Error calling HTTP GET for URL %s: %v", ovcli.ApiServerURL, err)
+		log.Error(err)
+		return nil, err
+	}
+	if len(body) == 0 {
+		err = fmt.Errorf("No system data found at %v", ovcli.ApiServerURL)
+		log.Error(err)
+		return nil, err
+	}
+
+	return ovcli, nil
+}
+
+// ensureAccessToken ensures that the access token exists
+func (o *Client) ensureAccessToken() error {
+	const tokenPath = "/sso/oauth/token"
+
+	if o.AccessToken != "" {
+		return nil
+	}
+
+	o.REST = ovhttp.NewRestClient(o, o.ApiServerURL, o.CA)
+
+	// create the payload to send using POST
+	d := url.Values{}
+	d.Set("username", o.Credentials.Username)
+	d.Set("password", o.Credentials.Password)
+	d.Set("scope", o.Credentials.Scope)
+	d.Set("grant_type", "password")
+
+	// call the server to get the access token
+	h := &http.Header{}
+	o.REST.HeaderUrlEncoded(h)
+	o.REST.HeaderAcceptJSON(h)
+	// path string, payload io.Reader, header *rest.Header
+	body, _, err := o.REST.Post(tokenPath, strings.NewReader(d.Encode()), h)
+	if err != nil {
+		err = fmt.Errorf("Error doing HTTP POST to oVirt server: %v", err)
+		log.Error(err)
+		return err
+	}
+	if len(body) == 0 {
+		err = fmt.Errorf("Error doing HTTP POST to create access token.  No body returned by server: %v", err)
+		log.Error(err)
+		return err
+	}
+
+	// extract access token from results
+	ad := &AuthData{}
+	if err = json.Unmarshal(body, ad); err != nil {
+		err = fmt.Errorf("Error UnMarshalling JSON for credentials: %v", err)
+		log.Error(err)
+		return err
+	}
+
+	o.AccessToken = ad.AccessToken
+	if o.AccessToken == "" {
+		err = fmt.Errorf("Access token missing from body returned by POST")
+		log.Error(err)
+		return err
+	}
+
+	return nil
+}
+
+// ClearAccessToken is called when there is a REST HTTP error
+func (o *Client) ClearAccessToken() {
+	o.AccessToken = ""
+}
+
+func getCredentials(cli kubernetes.Interface) (*Credentials, error) {
+	c := Credentials{}
+
+	c.Username = os.Getenv(olvm.EnvUsername)
+	if c.Username == "" {
```

Comment:
As a future enhancement, it would be nice if there was an attempt to prompt for these values if they are not specified and the command is being run in an interactive shell.

Reply:
let me create a JIRA